src/connector/src/parser/mod.rs

```diff
@@ -526,44 +669,20 @@ impl SpecificParserConfig {
         info: &StreamSourceInfo,
         props: &HashMap<String, String>,
     ) -> Result<Self> {
+        let parser_properties = ParserProperties::new(format, props, info)?;
         let conf = match format {
-            SourceFormat::Csv => SpecificParserConfig::Csv(CsvParserConfig {
-                delimiter: info.csv_delimiter as u8,
-                has_header: info.csv_has_header,
-            }),
-            SourceFormat::Avro => SpecificParserConfig::Avro(
-                AvroParserConfig::new(
-                    props,
-                    &info.row_schema_location,
-                    info.use_schema_registry,
-                    false,
-                    None,
-                )
-                .await?,
-            ),
-            SourceFormat::UpsertAvro => SpecificParserConfig::UpsertAvro(
-                AvroParserConfig::new(
-                    props,
-                    &info.row_schema_location,
-                    info.use_schema_registry,
-                    true,
-                    if info.upsert_avro_primary_key.is_empty() {
-                        None
-                    } else {
-                        Some(info.upsert_avro_primary_key.to_string())
-                    },
-                )
-                .await?,
-            ),
-            SourceFormat::Protobuf => SpecificParserConfig::Protobuf(
-                ProtobufParserConfig::new(
-                    props,
-                    &info.row_schema_location,
-                    &info.proto_message_name,
-                    info.use_schema_registry,
-                )
-                .await?,
-            ),
+            SourceFormat::Csv => {
+                SpecificParserConfig::Csv(CsvParserConfig::new(parser_properties)?)
+            }
+            SourceFormat::Avro => {
+                SpecificParserConfig::Avro(AvroParserConfig::new(parser_properties).await?)
+            }
+            SourceFormat::UpsertAvro => {
+                SpecificParserConfig::UpsertAvro(AvroParserConfig::new(parser_properties).await?)
+            }
+            SourceFormat::Protobuf => {
+                SpecificParserConfig::Protobuf(ProtobufParserConfig::new(parser_properties).await?)
+            }
```

Comment:
replace with macro here?

Reply:
I think we can wait until api is stable. I am not sure yet